options/options.go

```diff
@@ -795,5 +803,240 @@ type EngineOptions struct {
+	MaxAttempts      int
+	SleepIntervalSec int
+}
+
+// IgnoreConfig represents the configuration for ignoring specific errors
+type IgnoreConfig struct {
+	Name            string
+	IgnorableErrors []string
+	Message         string
+	Signals         map[string]interface{}
+}
+
+func cloneErrorsConfig(config *ErrorsConfig) *ErrorsConfig {
+	if config == nil {
+		return nil
+	}
+
+	// Create a new Errors
+	cloned := &ErrorsConfig{
+		Retry:  make(map[string]*RetryConfig),
+		Ignore: make(map[string]*IgnoreConfig),
+	}
+
+	// Clone Retry configurations
+	for key, retryConfig := range config.Retry {
+		if retryConfig != nil {
+			cloned.Retry[key] = &RetryConfig{
+				Name:             retryConfig.Name,
+				MaxAttempts:      retryConfig.MaxAttempts,
+				SleepIntervalSec: retryConfig.SleepIntervalSec,
+				RetryableErrors:  make([]string, len(retryConfig.RetryableErrors)),
+			}
+			// Deep copy the RetryableErrors slice
+			copy(cloned.Retry[key].RetryableErrors, retryConfig.RetryableErrors)
+		}
+	}
+
+	// Clone Ignore configurations
+	for key, ignoreConfig := range config.Ignore {
+		if ignoreConfig != nil {
+			cloned.Ignore[key] = &IgnoreConfig{
+				Name:            ignoreConfig.Name,
+				Message:         ignoreConfig.Message,
+				IgnorableErrors: make([]string, len(ignoreConfig.IgnorableErrors)),
+				Signals:         make(map[string]interface{}),
+			}
+			// Deep copy the IgnorableErrors slice
+			copy(cloned.Ignore[key].IgnorableErrors, ignoreConfig.IgnorableErrors)
+
+			// Deep copy the Signals map
+			for sigKey, sigVal := range ignoreConfig.Signals {
+				cloned.Ignore[key].Signals[sigKey] = sigVal
+			}
+		}
+	}
+
+	return cloned
+}
+
+// RunWithErrorHandling runs the given operation and handles any errors according to the configuration.
+func (opts *TerragruntOptions) RunWithErrorHandling(operation func() error) error {
+	if opts.Errors == nil {
+		return operation()
+	}
+
+	currentAttempt := 1
+
+	for {
+		err := operation()
+		if err == nil {
+			return nil
+		}
+
+		// Process the error through our error handling configuration
+		action, processErr := opts.Errors.ProcessError(err, currentAttempt)
+		if processErr != nil {
+			return fmt.Errorf("error processing error handling rules: %w", processErr)
+		}
+
+		if action == nil {
+			return err
+		}
+
+		if action.ShouldIgnore {
+			opts.Logger.Warnf("Ignoring error, reason: %s", action.IgnoreMessage)
+
+			// Handle ignore signals if any are configured
+			if len(action.IgnoreSignals) > 0 {
+				if err := opts.handleIgnoreSignals(action.IgnoreSignals); err != nil {
+					return err
+				}
+			}
+
+			return nil
+		}
+
+		if action.ShouldRetry {
+			opts.Logger.Warnf(
+				"Encountered retryable error: %s\nAttempt %d of %d. Waiting %d second(s) before retrying...",
+				action.RetryMessage,
+				currentAttempt,
+				action.RetryAttempts,
+				action.RetrySleepSecs,
+			)
+
+			// Sleep before retry
+			time.Sleep(time.Duration(action.RetrySleepSecs) * time.Second)
```

Comment:
```suggestion
			select {
			case <-time.After(time.Duration(action.RetrySleepSecs) * time.Second):
				// try again
			case <-ctx.Done():
				return errors.New(ctx.Err())
			}

```

Let's imagine that if an user sets RetrySleepSecs to say 60 seconds and RetryAttempts to 4 times and then wants to terminate the application by pressing ctrl-c, nothing will happen for 240 seconds. To prevent this, we need to listen to the context in case it was cancelled.